staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/envelope.go

```diff
@@ -117,22 +136,37 @@ func (t *envelopeTransformer) TransformFromStorage(data []byte, context value.Co
 
 // TransformToStorage encrypts data to be written to disk using envelope encryption.
 func (t *envelopeTransformer) TransformToStorage(data []byte, context value.Context) ([]byte, error) {
-	newKey, err := generateKey(32)
-	if err != nil {
-		return nil, err
-	}
-
-	encKey, err := t.envelopeService.Encrypt(newKey)
-	if err != nil {
-		// Do NOT wrap this err using fmt.Errorf() or similar functions
-		// because this gRPC status error has useful error code when
-		// record the metric.
-		return nil, err
-	}
+	var (
+		err         error
+		transformer value.Transformer
+	)
+	t.mux.Lock()
+	defer t.mux.Unlock()
+	// TODO(immutableT) Proactively generate a new DEK when the expiration period is
+	// approaching some threshold (ex. 15%). Fail-open before the expiration period, and
+	// instrument the failure, so that an alert could be generated.
+	if t.currentDEK.shouldRotate() {
+		t.currentDEK.plaintext, err = generateKey(32)
+		if err != nil {
+			return nil, err
+		}
+		t.currentDEK.ciphertext, err = t.envelopeService.Encrypt(t.currentDEK.plaintext)
+		if err != nil {
+			// Do NOT wrap this err using fmt.Errorf() or similar functions
+			// because this gRPC status error has useful error code when
+			// record the metric.
+			return nil, err
+		}
+		t.currentDEK.rotateAt = time.Now().Add(t.dekCacheTTL)
 
-	transformer, err := t.addTransformer(encKey, newKey)
-	if err != nil {
-		return nil, err
+		if transformer, err = t.addTransformer(t.currentDEK.ciphertext, t.currentDEK.plaintext); err != nil {
+			return nil, err
+		}
+	} else {
+		transformer = t.getTransformer(t.currentDEK.ciphertext)
```

Comment:
since it looks like you never clear t.currentDEK.plaintext, is this call needed?

Reply:
Removing this line bakes the unit test - it will panic in the scenario where the currentDEK is still fresh.

I am probably missing your point about "never clearing the t.currentDEK.plaintext", could you expand?
The plaintext is overwritten with a new value once the currentDEK expires.

Reply:
I mean that t.currentDEK.plaintext appears always to contain the plaintext DEK whose encrypted DEK is equal to t.currentDEK.ciphertext, so you don't need to get the transformer for t.currentDEK.ciphertext from the transformer cache --- you can just construct a new one from t.currentDEK.plaintext.

I think this might matter in the case where there is high secret *read* traffic, because t.currentDEK.ciphertext could indeed expire from the cache before its rotation period.

Consider: we have a TTL of, say, 5 minutes. I store one secret at the beginning of the rotation interval, then read 60*50 = 3000 secrets that happen to all have different DEKs over the next minute. That's more than the default cache size, so that t.currentDEK.ciphertext will have been wiped from the LRU cache. If I then try to write another secret, t.currentDEK.shouldRotate() will return false, but t.getTransformer(t.currentDEK.ciphertext) will fail.

Reply:
Thanks for catching this.
PTAL.